File: week_8_and_9/4_Ruby/oo_basics/my_solution.rb
# U3.W8-9: OO Basics: Student


# I worked on this challenge  with Devin

# 2. Pseudocode

# Create a class called students with attr accessor methods for scores,first_name and average
# Create a initializer method that takes arguments as name,scores 
# Set @first_name = name
# Set @scores = scores
# Set @average = total of elements of @scores array/length of @scores

# Create a method called letter_grade

# case @average
# when 90..100 return A
# when 80..89 return B
# when 70..79 return C
# when 60..69 return D
#else return E

# Create a empty array called students
# create a object called alex for Student class with arguments "Alex" and scores [100,100,100,0,100] and push it to students array
# create a object called sara for Student class with arguments "Sara" and scores [90,80,85,95,70] and push it to students array
# create a object called mike for Student class with arguments "Mike" and scores [75,90,68,55,90] and push it to students array
# create a object called mark for Student class with arguments "Mark" and scores [90,86,55,69,90] and push it to students array
# create a object called tim for Student class with arguments "Tim" and scores [90,99,100,76,87] and push it to students array

# Create a Method called linear_search that takes argument as student_list and name

# Loop through each element of array student_list
# if array element.first_name = name
# return the index of element
#else return -1
#end



# 3. Initial Solution
=begin
class Student
  attr_accessor :scores, :first_name

  def initialize(name,scores)   #Use named arguments!
    @first_name = name
    @scores = scores
  end
  
  def average
    avg = 0
    @scores.each do |x|
    avg += x
    end
    return avg / @scores.length
  end
  
  def letter_grade
    case self.average
    when 90..100
      "A"
    when 80..89
      "B"
    when 70..79
      "C"
    when 60..69
      "D"
    else
      "F"
    end
  end
  
end

students = []
students << alex = Student.new("Alex",[100,100,100,0,100])
students << sara = Student.new("Sara",[90,80,85,95,70])
students << mike = Student.new("Mike",[75,90,68,55,90])
students << mark = Student.new("Mark",[90,86,55,69,90])
students << tim = Student.new("Tim",[90,99,100,76,87])

def linear_search(student_list, name)
  student_list.each do |x|
    if x.first_name == name
      return student_list.index(x)
    end
  end
  return -1
end
=end
# 4. Refactored Solution

class Student
  attr_accessor :scores, :first_name, :average

  def initialize(name,scores)   #Use named arguments!
    @first_name = name
    @scores = scores
    @average = @scores.reduce(:+) / @scores.length
  end
  
  def letter_grade
    case @average
    when 90..100
      "A"
    when 80..89
      "B"
    when 70..79
      "C"
    when 60..69
      "D"
    else
      "F"
    end
  end
  
end

students = []
students << alex = Student.new("Alex",[100,100,100,0,100])
students << sara = Student.new("Sara",[90,80,85,95,70])
students << mike = Student.new("Mike",[75,90,68,55,90])
students << mark = Student.new("Mark",[90,86,55,69,90])
students << tim = Student.new("Tim",[90,99,100,76,87])

def linear_search(student_list, name)
  student_list.each do |student|
    if student.first_name == name
      return student_list.index(student)
    end
  end
  return -1
end





# 1. DRIVER TESTS GO BELOW THIS LINE
# Tests for release 0:

p students[0].first_name == "Alex"
p students[0].scores.length == 5
p students[0].scores[0] == students[0].scores[4]
p students[0].scores[3] == 0


# Tests for release 1:

p students[0].average == 80
p students[0].letter_grade == 'B'

# Tests for release 2:

p linear_search(students, "Alex") == 0
p linear_search(students, "NOT A STUDENT") == -1







# 5. Reflection 




# What parts of your strategy worked? What problems did you face?

#  Didnt face any problems as such. Had to think a bit on how to create the students array. 

# What questions did you have while coding? What resources did you find to help you answer them?

# Had to research on linear_search


# What concepts are you having trouble with, or did you just figure something out? If so, what?

# None



# Did you learn any new skills or tricks?

# linear_search


# How confident are you with each of the Learning Competencies?

# Very

# Which parts of the challenge did you enjoy?

# I paired with Devin and he showed me how to use case statements


# Which parts of the challenge did you find tedious?

# None

File: week_8_and_9/3_JavaScript/tally_votes/my_solution.js
// U3.W8-9: Gradebook from Names and Scores

// I worked on this challenge by myself

// These are the votes cast by each student. Do not alter these objects here.
var votes = {
  "Alex": { president: "Bob", vicePresident: "Devin", secretary: "Gail", treasurer: "Kerry" },
  "Bob": { president: "Mary", vicePresident: "Hermann", secretary: "Fred", treasurer: "Ivy" },
  "Cindy": { president: "Cindy", vicePresident: "Hermann", secretary: "Bob", treasurer: "Bob" },
  "Devin": { president: "Louise", vicePresident: "John", secretary: "Bob", treasurer: "Fred" },
  "Ernest": { president: "Fred", vicePresident: "Hermann", secretary: "Fred", treasurer: "Ivy" },
  "Fred": { president: "Louise", vicePresident: "Alex", secretary: "Ivy", treasurer: "Ivy" },
  "Gail": { president: "Fred", vicePresident: "Alex", secretary: "Ivy", treasurer: "Bob" },
  "Hermann": { president: "Ivy", vicePresident: "Kerry", secretary: "Fred", treasurer: "Ivy" },
  "Ivy": { president: "Louise", vicePresident: "Hermann", secretary: "Fred", treasurer: "Gail" },
  "John": { president: "Louise", vicePresident: "Hermann", secretary: "Fred", treasurer: "Kerry" },
  "Kerry": { president: "Fred", vicePresident: "Mary", secretary: "Fred", treasurer: "Ivy" },
  "Louise": { president: "Nate", vicePresident: "Alex", secretary: "Mary", treasurer: "Ivy" },
  "Mary": { president: "Louise", vicePresident: "Oscar", secretary: "Nate", treasurer: "Ivy" },
  "Nate": { president: "Oscar", vicePresident: "Hermann", secretary: "Fred", treasurer: "Tracy" },
  "Oscar": { president: "Paulina", vicePresident: "Nate", secretary: "Fred", treasurer: "Ivy" },
  "Paulina": { president: "Louise", vicePresident: "Bob", secretary: "Devin", treasurer: "Ivy" },
  "Quintin": { president: "Fred", vicePresident: "Hermann", secretary: "Fred", treasurer: "Bob" },
  "Romanda": { president: "Louise", vicePresident: "Steve", secretary: "Fred", treasurer: "Ivy" },
  "Steve": { president: "Tracy", vicePresident: "Kerry", secretary: "Oscar", treasurer: "Xavier" },
  "Tracy": { president: "Louise", vicePresident: "Hermann", secretary: "Fred", treasurer: "Ivy" },
  "Ullyses": { president: "Louise", vicePresident: "Hermann", secretary: "Ivy", treasurer: "Bob" },
  "Valorie": { president: "Wesley", vicePresident: "Bob", secretary: "Alex", treasurer: "Ivy" },
  "Wesley": { president: "Bob", vicePresident: "Yvonne", secretary: "Valorie", treasurer: "Ivy" },
  "Xavier": { president: "Steve", vicePresident: "Hermann", secretary: "Fred", treasurer: "Ivy" },
  "Yvonne": { president: "Bob", vicePresident: "Zane", secretary: "Fred", treasurer: "Hermann" },
  "Zane": { president: "Louise", vicePresident: "Hermann", secretary: "Fred", treasurer: "Mary" }
}

// Tally the votes in voteCount.
var voteCount = {
  president: {},
  vicePresident: {},
  secretary: {},
  treasurer: {}
}

/* The name of each student receiving a vote for an office should become a property 
of the respective office in voteCount.  After Alex's votes have been tallied, 
voteCount would be ...

  var voteCount = {
    president: { Bob: 1 },
    vicePresident: { Devin: 1 },
    secretary: { Gail: 1 },
    treasurer: { Kerry: 1 }
  }

*/


/* Once the votes have been tallied, assign each officer position the name of the 
student who received the most votes. */
var officers = {
  president: undefined,
  vicePresident: undefined,
  secretary: undefined,
  treasurer: undefined
}

// Pseudocode

// Loop through all the elements of the votes object
// for var names in votes
// assign var votes to value of name property
//  Loop through votes 
// Calculate the number of votes for each person for respective positions


    // Loop through the elements of the voteCount object
    // assign variable num to the values of voteCount property
    //Loop through num and set officers[office] to person with most counts
   

// __________________________________________
// Initial Solution

// for (var names in votes){
//           var vote = votes[names];
//           for (var tally in vote){
     
//       if (tally === "president"){
//       president = vote[tally];
      
//       if(voteCount.president[president] > 0)
//       {
//          voteCount.president[president] +=1;
//       }
//       else{
//       voteCount.president[president] =1;
// }
//       }
//           else if(tally === "vicePresident"){
//       vicepresident = vote[tally];
//       if( voteCount.vicePresident[vicepresident] > 0)
//       {
//          voteCount.vicePresident[vicepresident] += 1; 
//       }
//       else{
//       voteCount.vicePresident[vicepresident] = 1;
// }
//       }
//        else if(tally === "secretary"){
//       secretary = vote[tally];
//       if(voteCount.secretary[secretary] > 0){
//         voteCount.secretary[secretary] += 1;  
//       }
//       else{
//       voteCount.secretary[secretary] = 1;
// }
//       }
//           else if(tally === "treasurer"){
//       treasurer = vote[tally];
//       if(voteCount.treasurer[treasurer] > 0){
//              voteCount.treasurer[treasurer] += 1;

//       }
//       else{
//       voteCount.treasurer[treasurer] = 1;

//       }
//           }    
          
//       }
 
         
  
//       }
      



      
//       for (var officer in voteCount)
//       {
//           num = voteCount[officer];
//           var max = 1;
//           for(var a in num)
//           {
//              if (num[a] >=max)
//              {
//                  max = num[a];
                 
//                  officers[officer] = a;
//              }
            
    
//           }
//       }
      
      







// __________________________________________
// Refactored Solution

for (var names in votes){
  var vote = votes[names];
    for (var tally in vote){
      if(voteCount[tally][vote[tally]] > 0)
        {
           voteCount[tally][vote[tally]] +=1;
        }
      else{
           voteCount[tally][vote[tally]] = 1;
          }
          
    
    }
          
}

for (var officer in voteCount)
      {
          num = voteCount[officer];
          var max = 1;
          for(var a in num)
          {
             if (num[a] >=max)
             {
                 max = num[a];
                 
                 officers[officer] = a;
             }
            
    
          }
      }
      








// __________________________________________
// Driver Code:  Do not alter code below this line.


function assert(test, message, test_number) {
  if (!test) {
    console.log(test_number + "false");
    throw "ERROR: " + message;
  }
  console.log(test_number + "true");
  return true;
}

assert(
  (voteCount.president["Bob"] === 3),
  "Bob should receive three votes for President.",
  "1. "
)

assert(
  (voteCount.vicePresident["Bob"] === 2),
  "Bob should receive two votes for Vice President.",
  "2. "
)

assert(
  (voteCount.secretary["Bob"] === 2),
  "Bob should receive two votes for Secretary.",
  "3. "
)

assert(
  (voteCount.treasurer["Bob"] === 4),
  "Bob should receive four votes for Treasurer.",
  "4. "
)

assert(
  (officers.president === "Louise"),
  "Louise should be elected President.",
  "5. "
)

assert(
  (officers.vicePresident === "Hermann"),
  "Hermann should be elected Vice President.",
  "6. "
)

assert(
  (officers.secretary === "Fred"),
  "Fred should be elected Secretary.",
  "7. "
)

assert(
  (officers.treasurer === "Ivy"),
  "Ivy should be elected Treasurer.",
  "8. "
)


// __________________________________________
// Reflection

//What parts of your strategy worked? What problems did you face?

// Since I havent done similar exerices before it took me sometime to get it right. I had to try few different things before I could get it right.
//I used repl.it to test my code

// What questions did you have while coding? What resources did you find to help you answer them?

// I had to research on how to iterate over an objects property. I used the resources mentioned in the readme file


// What concepts are you having trouble with, or did you just figure something out? If so, what?

// Looping over objects property. Although I was able to figure it out, it seems a little more complicated than looping over arrays or hashes

// Did you learn any new skills or tricks?

// How to loop over javascript object property and accessing nested properties

// How confident are you with each of the Learning Competencies?

// Pretty confident. Though I need lot more practice

// Which parts of the challenge did you enjoy?

// Initially I was totally lost on this , but when I figured it out it was amazing.

// Which parts of the challenge did you find tedious?

// None.